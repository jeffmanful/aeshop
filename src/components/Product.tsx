import React, { useContext } from 'react'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'
import { ProductEntity } from '../types'
import { ReducerContext } from '../ProductsContext'
import likedIcon from '../liked.svg'
import unlikedIcon from '../unliked.svg'

const ProductWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 250px;
  padding: 1rem;
  transition: 0.4s all;

  p {
    font-weight: bold;
    margin: 5px;
  }

  img {
    cursor: pointer;
  }
`
ProductWrapper.displayName = 'Product'

const ProductImage = styled.img`
  max-width: 100%;
  cursor: pointer;
  transition: 0.3s all;

  &:hover {
    transform: translateY(-2px);
  }
`
ProductImage.displayName = 'ProductImage'

const ProductDetails = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-start;
`
ProductDetails.displayName = 'ProductDetails'

const Price = styled.p`
  font-size: 20px;
`
Price.displayName = 'Price'

const Sold = styled.p`
  color: red;
`
Sold.displayName = 'Sold'

const Product: React.FC<ProductEntity> = ({
  title,
  description,
  sold,
  img,
  price,
  id,
  liked,
  size,
  brand
}): JSX.Element => {
  const dispatch = useContext(ReducerContext)
  const history = useHistory()

  const formattedPrice = `Â£${price}`
  const currentIcon = liked ? likedIcon : unlikedIcon
  return (
    <ProductWrapper data-testid="product">
      <ProductImage
        src={img}
        alt={description}
        onClick={() => history.push(`/product/${id}`)}
      />
      <div
        className="row"
        style={{ display: 'flex', justifyContent: 'space-between' }}
      >
        <ProductDetails>
          <p>{title}</p>
          <p>{size}</p>
          <p>{brand}</p>
          {sold ? <Sold>SOLD</Sold> : <Price>{formattedPrice}</Price>}
        </ProductDetails>
        <img
          data-testid={currentIcon}
          role="button"
          src={currentIcon}
          alt="like icon"
          onClick={() => dispatch({ type: 'TOGGLE_PRODUCT_LIKED', id })}
        />
      </div>
    </ProductWrapper>
  )
}

export default Product
