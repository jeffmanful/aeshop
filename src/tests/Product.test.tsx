import React from 'react'
import Product from '../components/Product'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { render, fireEvent } from '@testing-library/react'

describe('ProductTest', () => {
  const history = createMemoryHistory()
  const baseProduct = {
    title: 'title',
    description: '',
    sold: false,
    img: '',
    price: '22.22',
    id: 1,
    liked: false,
    seller: '',
    slug: '',
  }

  const { asFragment, queryByText } = render(
    <Router history={history}>
      <Product {...baseProduct} />
    </Router>
  )

  test('it renders', () => {
    expect(queryByText('title')).toBeInTheDocument()
    expect(queryByText('Â£22.22')).toBeInTheDocument()

    expect(queryByText('SOLD')).toBe(null)
    expect(asFragment()).toMatchSnapshot()
  })

  test('like product', () => {
    const { getByTestId, getByRole } = render(
      <Router history={history}>
        <Product {...baseProduct} />
      </Router>
    )
    expect(getByTestId('unliked.svg')).toBeInTheDocument()

    fireEvent.click(getByRole('button'))

    expect(getByTestId(/liked.svg/)).toBeInTheDocument()
  })

  test('sold product', () => {
    const soldProduct = { ...baseProduct, sold: true }
    const { queryByText } = render(
      <Router history={history}>
        <Product {...soldProduct} />
      </Router>
    )

    expect(queryByText('SOLD')).toBeInTheDocument()
  })
})
