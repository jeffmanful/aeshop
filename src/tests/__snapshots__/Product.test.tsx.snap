// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTest it renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm iHyQVR"
    data-testid="product"
  >
    <img
      alt=""
      class="sc-AxirZ dVmuFx"
      src=""
    />
    <div
      class="row"
      style="display: flex; justify-content: space-between;"
    >
      <div
        class="sc-AxiKw clJdzB"
      >
        <p>
          title
        </p>
        <p />
        <p />
        <p
          class="sc-AxhCb gjTtwG"
        >
          Â£22.22
        </p>
      </div>
      <img
        alt="like icon"
        data-testid="unliked.svg"
        role="button"
        src="unliked.svg"
      />
    </div>
  </div>
</DocumentFragment>
`;
